File: sdk/ts/src/auth/workload/gcp.ts
/** GCP workload identity source via the metadata server. */

import { TOKEN_TYPE_ID_TOKEN } from "../exchange.js";
import type { SubjectToken, WorkloadIdentitySource } from "./index.js";

export const METADATA_HOST = "metadata.google.internal";
export const IDENTITY_PATH = "/computeMetadata/v1/instance/service-accounts/default/identity";

export class GCPWorkloadIdentitySource implements WorkloadIdentitySource {
  readonly name = "gcp";
  private readonly audience: string | undefined;
  private readonly timeoutMs: number;
  private readonly fetchImpl: typeof fetch;

  constructor(opts: { audience?: string; timeoutMs?: number; fetchImpl?: typeof fetch } = {}) {
    this.audience = opts.audience;
    this.timeoutMs = opts.timeoutMs ?? 2_000;
    this.fetchImpl = opts.fetchImpl ?? fetch;
  }

  async fetch(): Promise<SubjectToken | null> {
    if (!looksLikeGCP()) return null;

    const audience = this.audience ?? processEnv().MARMOT_HOST;
    if (!audience) return null;

    const url = new URL(`http://${METADATA_HOST}${IDENTITY_PATH}`);
    url.searchParams.set("audience", audience);
    url.searchParams.set("format", "full");

    const ac = new AbortController();
    const timer = setTimeout(() => ac.abort(), this.timeoutMs);
    try {
      const resp = await this.fetchImpl(url, {
        headers: { "Metadata-Flavor": "Google" },
        signal: ac.signal,
      });
      if (!resp.ok) return null;
      const text = (await resp.text()).trim();
      if (!text) return null;
      return { token: text, tokenType: TOKEN_TYPE_ID_TOKEN };
    } catch {
      return null;
    } finally {
      clearTimeout(timer);
    }
  }
}

function looksLikeGCP(): boolean {
  const env = processEnv();
  return Boolean(
    env.GOOGLE_CLOUD_PROJECT || env.GCLOUD_PROJECT || env.K_SERVICE || env.FUNCTION_TARGET,
  );
}

function processEnv(): Record<string, string | undefined> {
  return typeof process !== "undefined" && process.env ? process.env : {};
}

File: sdk/python/src/marmot/auth/workload/gcp.py
"""GCP workload identity source.

On GCE / GKE / Cloud Run / Cloud Functions, Google's metadata server vends an
identity token for the active service account when queried with the right
audience. We pass the Marmot host as the audience so the token can only be
replayed against this server.
"""

from __future__ import annotations

import os
from dataclasses import dataclass

import httpx

from marmot.auth.workload import SubjectToken, register_source
from marmot.config import ENVIRONMENT_HOST

METADATA_HOST = "metadata.google.internal"
IDENTITY_PATH = "/computeMetadata/v1/instance/service-accounts/default/identity"
METADATA_HEADER = {"Metadata-Flavor": "Google"}


@register_source
@dataclass
class GCPWorkloadIdentitySource:
    name: str = "gcp"
    audience: str | None = None  # defaults to the Marmot host at fetch time
    timeout: float = 2.0

    def fetch(self, audience: str | None = None) -> SubjectToken | None:
        if not _looks_like_gcp():
            return None

        audience = self.audience or audience or os.environ.get(ENVIRONMENT_HOST)
        if not audience:
            return None

        try:
            resp = httpx.get(
                f"http://{METADATA_HOST}{IDENTITY_PATH}",
                params={"audience": audience, "format": "full"},
                headers=METADATA_HEADER,
                timeout=self.timeout,
            )
        except httpx.HTTPError:
            return None

        if resp.status_code != 200:
            return None

        token = resp.text.strip()
        if not token:
            return None
        return SubjectToken(token=token, source=self.name)


def _looks_like_gcp() -> bool:
    return any(
        var in os.environ
        for var in (
            "GOOGLE_CLOUD_PROJECT",
            "GCLOUD_PROJECT",
            "K_SERVICE",  # Cloud Run / Cloud Functions
            "FUNCTION_TARGET",
        )
    )
